crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -420,7 +421,11 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
             TimelineEventKind::UnableToDecrypt { content, unable_to_decrypt_info } => {
                 // TODO: Handle replacements if the replaced event is also UTD
                 let raw_event = self.ctx.flow.raw_event();
-                let cause = UtdCause::determine(raw_event, &unable_to_decrypt_info);
+                let client_info = match self.meta.unable_to_decrypt_hook.as_ref() {
+                    Some(hook) => Some(hook.client_info().await),
+                    None => None,
+                };
```

Comment:
This doesn't belong in the UTD hook. (For example: the `cause` passed to `TimelineItemContent::unable_to_decrypt` should not depend on whether a UTD hook has been registered or not). We need to find another way to access the `ClientInfo` here.

How about:
 * Add a `client_info` method to `RoomDataProvider` (it is implemented by `Room`, which has a reference to the Client, so that should be easy)
 * Add a `UtdCause` filed to `matrix_sdk_ui::timeline::event_handler::TimelineEventKind::UnableToDecrypt`
 * Change `TimelineEventKind::from_event` to take a `RoomDataProvider` instead of a `RoomVersionId`
 * Call `UtdCause::determine()` in `TimelineEventKind::from_event`, to populate the UtdCause

Reply:
(incidentally, we may want to think of a more descriptive name than "ClientInfo" here)

Reply:
I have implemented some changes, better to look at commit per commit
- https://github.com/matrix-org/matrix-rust-sdk/pull/4275/commits/db1fadeb51f72fe7aeededdce389e4e0cc96b364 Moved utd_cause to the sdk-ui UnableToDecrypt kind
- https://github.com/matrix-org/matrix-rust-sdk/pull/4275/commits/a4092232a49375577b9a11cf5febbc4946399957 As discussed the unable_to_decrypt_info is not needed in the ui crate, the utd_cause is enough
- https://github.com/matrix-org/matrix-rust-sdk/pull/4275/commits/0538bcac347a9013e543b646de7eb5e502771b10 Rename client info
- https://github.com/matrix-org/matrix-rust-sdk/pull/4275/commits/749c0d1a018fcf05efcc9bc484f227ce48a15d11 the e2e feature flag is not visible in the UI crate, so made the API returns an optional `CryptoContextInfo` and return None in sdk if feature not enabled